cpp/include/Ice/ServerAuthenticationOptions.h

```diff
@@ -0,0 +1,161 @@
+         * set it has preference over a certificate chain set in serverCertificateChain.
+         *
+         * @param host The server's host name.
+         * @return The server's certificate chain.
+         */
+        std::function<CFArrayRef(const std::string& host)> serverCertificateSelectionCallback;
+
+        /**
+         * The trusted root certificates. If set, the client's certificate chain is validated against these
+         * certificates. If not set the system's root certificates are used.
+         */
+        CFArrayRef trustedRootCertificates;
+
+        /**
+         * A callback that allows manually validating the client certificate chain.
+         *
+         * The validation callback should return true if the certificate chain is valid, false otherwise. Alternatively
+         * an exception can be throw to indicate an error.
+         *
+         * @param trust The trust object that contains the client's certificate chain.
+         * @param info The connection information.
+         * @return true if the certificate chain is valid, false otherwise.
+         */
+        std::function<bool(SecTrustRef trust, const IceSSL::ConnectionInfoPtr& info)>
+            clientCertificateValidationCallback;
+
+        /**
+         * the requirements for client-side authentication
+         * @see https://developer.apple.com/documentation/security/sslauthenticate
+         */
+        SSLAuthenticate clientCertificateRequired;
+
+        /**
+         * A callback that is called before a new SSL handshake starts. The callback can be used to set additional SSL
+         * context parameters.
+         */
+        std::function<void(SSLContextRef)> sslNewSessionCallback;
+#else
+        /**
+         * The server's SSL context, this objects holds configuration relevant the SSL session establishment.
+         */
+        SSL_CTX* serverSslContext;
+
+        /**
+         * A callback that is called before a new SSL handshake starts. The callback can be used to set additional SSL
+         * parameters.
+         */
+        std::function<void(::SSL* ssl, const std::string& host)> sslNewSessionCallback;
+
+        /**
+         * A callback that allows manually validating the client certificate chain.
+         *
+         * The validation callback should return true if the certificate chain is valid, false otherwise. Alternatively
+         * an exception can be throw to indicate an error.
+         *
+         * @param ok Whether the verification of the certificate in question was passed (ok=1) or not (ok=0).
+         * @param ctx The X509_STORE_CTX object that holds the certificate chain to be verified.
+         * @param info The connection information.
+         * @return true if the certificate chain is valid, false otherwise.
+         */
+        std::function<int(int ok, X509_STORE_CTX*, const IceSSL::ConnectionInfoPtr& info)>
+            clientCertificateVerificationCallback;
```

Comment:
The implementation wraps this callback and uses it as the OpenSSL verification callback. The signature is the same with the additional `ConnectionInfo`. The info param is required for the trust verifier, but I plan to remove this dependency in a follow-up PR.

The wrapper saves exceptions thrown from the callback, and rethrows them from the transceiver initialization.